image/manifest.go

```diff
@@ -240,3 +240,33 @@ loop:
 	}
 	return nil
 }
+
+// ValidateManifestMediaType validate the manifest schema media-type
+// and its descendants fields media type, such as config and layers,
+// to make sure they match to spec definition, or returns an error if
+// the validation failed.
+func ValidateManifestMediaType(r io.Reader) error {
+	header := v1.Manifest{}
+
+	buf, err := ioutil.ReadAll(r)
+	if err != nil {
+		return errors.Wrapf(err, "error reading the io stream")
+	}
+
+	err = json.Unmarshal(buf, &header)
```

Comment:
There is still no reason to read `r` into a buffer and then unpack it to JSON.  You've previously used:

    buf, err := ioutil.ReadAll(r)
    …
    err = json.NewDecoder(bytes.NewReader(buf)).Decode(&header)

And we all agreed that the `NewReader` call wasn't a good idea.  But instead of `Decode` → `Unmarshal`, I [still think][1] you just want to use:

    func ValidateManifestMediaType(reader io.Reader) error {
      manifest := v1.Manifest{}
      err = json.NewDecoder(reader).Decode(&manifest)
      …
[1]: https://github.com/opencontainers/image-tools/pull/10/#discussion_r79662962